File: src/formats/ModelCfg.cpp
#include "ModelCfg.h"

#include <algorithm>
#include <cctype>
#include <fstream>
#include <sstream>
#include <stdexcept>

namespace a3ob::cfg
{
namespace
{
enum class TokenKind
{
    Identifier,
    String,
    Integer,
    Number,
    Class,
    BraceOpen,
    BraceClose,
    BracketOpen,
    BracketClose,
    Equals,
    PlusEquals,
    Semicolon,
    Comma,
    Colon,
    End,
};

struct Token
{
    TokenKind kind = TokenKind::End;
    std::string text;
};

std::string lower(std::string value)
{
    std::transform(value.begin(), value.end(), value.begin(), [](unsigned char ch) { return static_cast<char>(std::tolower(ch)); });
    return value;
}

std::string readTextFile(const std::filesystem::path& path)
{
    std::ifstream stream(path, std::ios::binary);
    if (!stream) {
        throw std::runtime_error("Could not open model.cfg: " + path.string());
    }
    std::ostringstream buffer;
    buffer << stream.rdbuf();
    return buffer.str();
}

std::vector<Token> tokenize(const std::string& text)
{
    std::vector<Token> tokens;
    std::size_t i = 0;
    while (i < text.size()) {
        const char ch = text[i];
        if (std::isspace(static_cast<unsigned char>(ch))) {
            ++i;
            continue;
        }
        if (ch == '/' && i + 1 < text.size() && text[i + 1] == '/') {
            i += 2;
            while (i < text.size() && text[i] != '\n') ++i;
            continue;
        }
        if (ch == '/' && i + 1 < text.size() && text[i + 1] == '*') {
            i += 2;
            while (i + 1 < text.size() && !(text[i] == '*' && text[i + 1] == '/')) ++i;
            i = std::min(i + 2, text.size());
            continue;
        }
        if (ch == '#') {
            throw std::runtime_error("model.cfg preprocessor directives are not supported");
        }
        if (std::isalpha(static_cast<unsigned char>(ch)) || ch == '_') {
            std::size_t start = i++;
            while (i < text.size() && (std::isalnum(static_cast<unsigned char>(text[i])) || text[i] == '_')) ++i;
            std::string value = text.substr(start, i - start);
            tokens.push_back({lower(value) == "class" ? TokenKind::Class : TokenKind::Identifier, value});
            continue;
        }
        if (std::isdigit(static_cast<unsigned char>(ch)) || ((ch == '-' || ch == '+') && i + 1 < text.size() && std::isdigit(static_cast<unsigned char>(text[i + 1])))) {
            std::size_t start = i++;
            bool isFloat = false;
            while (i < text.size() && (std::isdigit(static_cast<unsigned char>(text[i])) || text[i] == '.')) {
                if (text[i] == '.') isFloat = true;
                ++i;
            }
            tokens.push_back({isFloat ? TokenKind::Number : TokenKind::Integer, text.substr(start, i - start)});
            continue;
        }
        if (ch == '"') {
            ++i;
            std::string value;
            while (i < text.size()) {
                if (text[i] == '"') {
                    if (i + 1 < text.size() && text[i + 1] == '"') {
                        value.push_back('"');
                        i += 2;
                        continue;
                    }
                    ++i;
                    break;
                }
                value.push_back(text[i++]);
            }
            tokens.push_back({TokenKind::String, value});
            continue;
        }
        switch (ch) {
        case '{': tokens.push_back({TokenKind::BraceOpen, "{"}); ++i; break;
        case '}': tokens.push_back({TokenKind::BraceClose, "}"}); ++i; break;
        case '[': tokens.push_back({TokenKind::BracketOpen, "["}); ++i; break;
        case ']': tokens.push_back({TokenKind::BracketClose, "]"}); ++i; break;
        case '=': tokens.push_back({TokenKind::Equals, "="}); ++i; break;
        case '+':
            if (i + 1 < text.size() && text[i + 1] == '=') {
                tokens.push_back({TokenKind::PlusEquals, "+="});
                i += 2;
            } else {
                throw std::runtime_error("Unexpected plus token in model.cfg");
            }
            break;
        case ';': tokens.push_back({TokenKind::Semicolon, ";"}); ++i; break;
        case ',': tokens.push_back({TokenKind::Comma, ","}); ++i; break;
        case ':': tokens.push_back({TokenKind::Colon, ":"}); ++i; break;
        default:
            throw std::runtime_error(std::string("Unexpected character in model.cfg: ") + ch);
        }
    }
    tokens.push_back({TokenKind::End, {}});
    return tokens;
}

class Parser
{
public:
    explicit Parser(std::vector<Token> tokens) : tokens_(std::move(tokens)) {}

    Class parseRoot()
    {
        Class root;
        root.name = "__root__";
        while (!check(TokenKind::End)) {
            root.classes.push_back(parseClass());
        }
        return root;
    }

private:
    bool check(TokenKind kind) const { return tokens_[pos_].kind == kind; }

    const Token& consume(TokenKind kind, const char* message)
    {
        if (!check(kind)) {
            throw std::runtime_error(message);
        }
        return tokens_[pos_++];
    }

    Class parseClass()
    {
        consume(TokenKind::Class, "Expected class keyword");
        Class cls;
        cls.name = consume(TokenKind::Identifier, "Expected class name").text;
        if (check(TokenKind::Semicolon)) {
            ++pos_;
            return cls;
        }
        if (check(TokenKind::Colon)) {
            ++pos_;
            cls.parent = consume(TokenKind::Identifier, "Expected parent class name").text;
        }
        consume(TokenKind::BraceOpen, "Expected class body");
        while (!check(TokenKind::BraceClose)) {
            if (check(TokenKind::Class)) {
                cls.classes.push_back(parseClass());
            } else {
                cls.properties.push_back(parseProperty());
            }
        }
        consume(TokenKind::BraceClose, "Expected class closing brace");
        consume(TokenKind::Semicolon, "Expected semicolon after class");
        return cls;
    }

    Property parseProperty()
    {
        Property property;
        property.name = consume(TokenKind::Identifier, "Expected property name").text;
        if (check(TokenKind::BracketOpen)) {
            ++pos_;
            consume(TokenKind::BracketClose, "Expected [] property close");
        }
        if (check(TokenKind::PlusEquals)) {
            property.extends = true;
            ++pos_;
        } else {
            consume(TokenKind::Equals, "Expected property assignment");
        }
        property.value = parseValue();
        consume(TokenKind::Semicolon, "Expected semicolon after property");
        return property;
    }

    Value parseValue()
    {
        if (check(TokenKind::BraceOpen)) {
            ++pos_;
            Array values;
            while (!check(TokenKind::BraceClose)) {
                values.push_back(parseValue());
                if (check(TokenKind::Comma)) ++pos_;
            }
            consume(TokenKind::BraceClose, "Expected array closing brace");
            return Value{values};
        }
        if (check(TokenKind::String)) return Value{tokens_[pos_++].text};
        if (check(TokenKind::Integer)) return Value{std::stoi(tokens_[pos_++].text)};
        if (check(TokenKind::Number)) return Value{std::stod(tokens_[pos_++].text)};
        if (check(TokenKind::Identifier)) return Value{tokens_[pos_++].text};
        throw std::runtime_error("Expected config value");
    }

    std::vector<Token> tokens_;
    std::size_t pos_ = 0;
};

void writeIndent(std::ostream& stream, int indent)
{
    for (int i = 0; i < indent; ++i) stream << '\t';
}

void writeValue(std::ostream& stream, const Value& value, int indent);

void writeArray(std::ostream& stream, const Array& array, int indent)
{
    if (array.empty()) {
        stream << "{}";
        return;
    }
    stream << "{\n";
    for (std::size_t i = 0; i < array.size(); ++i) {
        writeIndent(stream, indent + 1);
        writeValue(stream, array[i], indent + 1);
        if (i + 1 < array.size()) stream << ',';
        stream << "\n";
    }
    writeIndent(stream, indent);
    stream << '}';
}

void writeValue(std::ostream& stream, const Value& value, int indent)
{
    if (value.isString()) {
        stream << '"' << value.asString() << '"';
    } else if (value.isInt()) {
        stream << value.asInt();
    } else if (value.isDouble()) {
        stream << value.asDouble();
    } else {
        writeArray(stream, value.asArray(), indent);
    }
}

void writeClass(std::ostream& stream, const Class& cls, int indent)
{
    writeIndent(stream, indent);
    stream << "class " << cls.name;
    if (!cls.parent.empty()) stream << ": " << cls.parent;
    stream << " {\n";
    for (const Property& property : cls.properties) {
        writeIndent(stream, indent + 1);
        stream << property.name;
        if (property.value.isArray()) stream << "[]";
        stream << (property.extends ? " += " : " = ");
        writeValue(stream, property.value, indent + 1);
        stream << ";\n";
    }
    for (const Class& child : cls.classes) {
        writeClass(stream, child, indent + 1);
    }
    writeIndent(stream, indent);
    stream << "};\n";
}
}

bool Value::isString() const { return std::holds_alternative<std::string>(data); }
bool Value::isInt() const { return std::holds_alternative<int>(data); }
bool Value::isDouble() const { return std::holds_alternative<double>(data); }
bool Value::isArray() const { return std::holds_alternative<Array>(data); }
const std::string& Value::asString() const { return std::get<std::string>(data); }
int Value::asInt() const { return std::get<int>(data); }
double Value::asDouble() const { return std::get<double>(data); }
const Array& Value::asArray() const { return std::get<Array>(data); }

const Class* Class::findClass(const std::string& className) const
{
    const std::string needle = lower(className);
    for (const Class& child : classes) {
        if (lower(child.name) == needle) return &child;
    }
    return nullptr;
}

const Property* Class::findProperty(const std::string& propertyName) const
{
    const std::string needle = lower(propertyName);
    for (const Property& property : properties) {
        if (lower(property.name) == needle) return &property;
    }
    return nullptr;
}

Config Config::readFile(const std::filesystem::path& path)
{
    Parser parser(tokenize(readTextFile(path)));
    Config config;
    config.root = parser.parseRoot();
    return config;
}

void Config::writeFile(const std::filesystem::path& path) const
{
    std::ofstream stream(path, std::ios::binary);
    if (!stream) {
        throw std::runtime_error("Could not write model.cfg: " + path.string());
    }
    for (const Class& cls : root.classes) {
        writeClass(stream, cls, 0);
        stream << '\n';
    }
}

std::vector<Skeleton> Config::skeletons() const
{
    std::vector<Skeleton> result;
    const Class* cfgSkeletons = root.findClass("CfgSkeletons");
    if (!cfgSkeletons) return result;
    for (const Class& skeletonClass : cfgSkeletons->classes) {
        Skeleton skeleton;
        skeleton.name = skeletonClass.name;
        const Property* bones = skeletonClass.findProperty("skeletonBones");
        if (bones && bones->value.isArray()) {
            const Array& values = bones->value.asArray();
            for (std::size_t i = 0; i + 1 < values.size(); i += 2) {
                if (values[i].isString() && values[i + 1].isString()) {
                    skeleton.bones.push_back({values[i].asString(), values[i + 1].asString()});
                }
            }
        }
        result.push_back(std::move(skeleton));
    }
    return result;
}

Config Config::skeletonConfig(const Skeleton& skeleton)
{
    Class skeletonClass;
    skeletonClass.name = skeleton.name;
    skeletonClass.properties.push_back({"isDiscrete", Value{0}, false});
    skeletonClass.properties.push_back({"skeletonInherit", Value{std::string()}, false});
    Array bones;
    for (const SkeletonBone& bone : skeleton.bones) {
        bones.push_back(Value{bone.name});
        bones.push_back(Value{bone.parent});
    }
    skeletonClass.properties.push_back({"skeletonBones", Value{bones}, false});

    Class cfgSkeletons;
    cfgSkeletons.name = "CfgSkeletons";
    cfgSkeletons.classes.push_back(std::move(skeletonClass));

    Config config;
    config.root.name = "__root__";
    config.root.classes.push_back(std::move(cfgSkeletons));
    return config;
}
}

File: scripts/a3ob/formats/model_cfg.py
"""Bohemia ``model.cfg`` / class-config parser and writer.

Direct 1:1 port of the former C++ ``src/formats/ModelCfg.{h,cpp}``. A small
recursive-descent parser over the class-config syntax used by ``CfgSkeletons``.
Preprocessor directives (``#include`` etc.) are not supported and raise.

Values are plain Python ``str`` / ``int`` / ``float`` / ``list`` (arrays). Bareword
identifiers parse as strings and are re-quoted on write, matching the C++ behavior.
"""

import io

__all__ = [
    "Value", "Property", "Class", "SkeletonBone", "Skeleton", "Config",
]


# -- tokenizer -----------------------------------------------------------------
_IDENTIFIER = "Identifier"
_STRING = "String"
_INTEGER = "Integer"
_NUMBER = "Number"
_CLASS = "Class"
_BRACE_OPEN = "BraceOpen"
_BRACE_CLOSE = "BraceClose"
_BRACKET_OPEN = "BracketOpen"
_BRACKET_CLOSE = "BracketClose"
_EQUALS = "Equals"
_PLUS_EQUALS = "PlusEquals"
_SEMICOLON = "Semicolon"
_COMMA = "Comma"
_COLON = "Colon"
_END = "End"


def _tokenize(text):
    tokens = []
    i = 0
    n = len(text)
    while i < n:
        ch = text[i]
        if ch.isspace():
            i += 1
            continue
        if ch == "/" and i + 1 < n and text[i + 1] == "/":
            i += 2
            while i < n and text[i] != "\n":
                i += 1
            continue
        if ch == "/" and i + 1 < n and text[i + 1] == "*":
            i += 2
            while i + 1 < n and not (text[i] == "*" and text[i + 1] == "/"):
                i += 1
            i = min(i + 2, n)
            continue
        if ch == "#":
            raise ValueError("model.cfg preprocessor directives are not supported")
        if ch.isalpha() or ch == "_":
            start = i
            i += 1
            while i < n and (text[i].isalnum() or text[i] == "_"):
                i += 1
            value = text[start:i]
            kind = _CLASS if value.lower() == "class" else _IDENTIFIER
            tokens.append((kind, value))
            continue
        if ch.isdigit() or ((ch == "-" or ch == "+") and i + 1 < n and text[i + 1].isdigit()):
            start = i
            i += 1
            is_float = False
            while i < n and (text[i].isdigit() or text[i] == "."):
                if text[i] == ".":
                    is_float = True
                i += 1
            tokens.append((_NUMBER if is_float else _INTEGER, text[start:i]))
            continue
        if ch == '"':
            i += 1
            chars = []
            while i < n:
                if text[i] == '"':
                    if i + 1 < n and text[i + 1] == '"':
                        chars.append('"')
                        i += 2
                        continue
                    i += 1
                    break
                chars.append(text[i])
                i += 1
            tokens.append((_STRING, "".join(chars)))
            continue

        simple = {
            "{": _BRACE_OPEN, "}": _BRACE_CLOSE,
            "[": _BRACKET_OPEN, "]": _BRACKET_CLOSE,
            "=": _EQUALS, ";": _SEMICOLON, ",": _COMMA, ":": _COLON,
        }
        if ch in simple:
            tokens.append((simple[ch], ch))
            i += 1
            continue
        if ch == "+":
            if i + 1 < n and text[i + 1] == "=":
                tokens.append((_PLUS_EQUALS, "+="))
                i += 2
                continue
            raise ValueError("Unexpected plus token in model.cfg")
        raise ValueError("Unexpected character in model.cfg: " + ch)

    tokens.append((_END, ""))
    return tokens


# -- data model ----------------------------------------------------------------
class Value:
    __slots__ = ("data",)

    def __init__(self, data):
        self.data = data

    def is_string(self):
        return isinstance(self.data, str)

    def is_int(self):
        return isinstance(self.data, int) and not isinstance(self.data, bool)

    def is_double(self):
        return isinstance(self.data, float)

    def is_array(self):
        return isinstance(self.data, list)

    def as_string(self):
        return self.data

    def as_int(self):
        return self.data

    def as_double(self):
        return self.data

    def as_array(self):
        return self.data


class Property:
    __slots__ = ("name", "value", "extends")

    def __init__(self, name="", value=None, extends=False):
        self.name = name
        self.value = value
        self.extends = extends


class Class:
    __slots__ = ("name", "parent", "properties", "classes")

    def __init__(self, name="", parent=""):
        self.name = name
        self.parent = parent
        self.properties = []
        self.classes = []

    def find_class(self, class_name):
        needle = class_name.lower()
        for child in self.classes:
            if child.name.lower() == needle:
                return child
        return None

    def find_property(self, property_name):
        needle = property_name.lower()
        for prop in self.properties:
            if prop.name.lower() == needle:
                return prop
        return None


class SkeletonBone:
    __slots__ = ("name", "parent")

    def __init__(self, name="", parent=""):
        self.name = name
        self.parent = parent


class Skeleton:
    __slots__ = ("name", "bones")

    def __init__(self, name=""):
        self.name = name
        self.bones = []


# -- parser --------------------------------------------------------------------
class _Parser:
    def __init__(self, tokens):
        self._tokens = tokens
        self._pos = 0

    def parse_root(self):
        root = Class("__root__")
        while not self._check(_END):
            root.classes.append(self._parse_class())
        return root

    def _check(self, kind):
        return self._tokens[self._pos][0] == kind

    def _consume(self, kind, message):
        if not self._check(kind):
            raise ValueError(message)
        token = self._tokens[self._pos]
        self._pos += 1
        return token

    def _parse_class(self):
        self._consume(_CLASS, "Expected class keyword")
        cls = Class()
        cls.name = self._consume(_IDENTIFIER, "Expected class name")[1]
        if self._check(_SEMICOLON):
            self._pos += 1
            return cls
        if self._check(_COLON):
            self._pos += 1
            cls.parent = self._consume(_IDENTIFIER, "Expected parent class name")[1]
        self._consume(_BRACE_OPEN, "Expected class body")
        while not self._check(_BRACE_CLOSE):
            if self._check(_CLASS):
                cls.classes.append(self._parse_class())
            else:
                cls.properties.append(self._parse_property())
        self._consume(_BRACE_CLOSE, "Expected class closing brace")
        self._consume(_SEMICOLON, "Expected semicolon after class")
        return cls

    def _parse_property(self):
        prop = Property()
        prop.name = self._consume(_IDENTIFIER, "Expected property name")[1]
        if self._check(_BRACKET_OPEN):
            self._pos += 1
            self._consume(_BRACKET_CLOSE, "Expected [] property close")
        if self._check(_PLUS_EQUALS):
            prop.extends = True
            self._pos += 1
        else:
            self._consume(_EQUALS, "Expected property assignment")
        prop.value = self._parse_value()
        self._consume(_SEMICOLON, "Expected semicolon after property")
        return prop

    def _parse_value(self):
        if self._check(_BRACE_OPEN):
            self._pos += 1
            values = []
            while not self._check(_BRACE_CLOSE):
                values.append(self._parse_value())
                if self._check(_COMMA):
                    self._pos += 1
            self._consume(_BRACE_CLOSE, "Expected array closing brace")
            return Value(values)
        if self._check(_STRING):
            return Value(self._tokens[self._pos_advance()][1])
        if self._check(_INTEGER):
            return Value(int(self._tokens[self._pos_advance()][1]))
        if self._check(_NUMBER):
            return Value(float(self._tokens[self._pos_advance()][1]))
        if self._check(_IDENTIFIER):
            return Value(self._tokens[self._pos_advance()][1])
        raise ValueError("Expected config value")

    def _pos_advance(self):
        pos = self._pos
        self._pos += 1
        return pos


# -- writer --------------------------------------------------------------------
def _write_indent(stream, indent):
    stream.write("\t" * indent)


def _write_value(stream, value, indent):
    if value.is_string():
        stream.write('"' + value.as_string() + '"')
    elif value.is_int():
        stream.write(str(value.as_int()))
    elif value.is_double():
        stream.write(_format_double(value.as_double()))
    else:
        _write_array(stream, value.as_array(), indent)


def _format_double(value):
    # The C++ writer used default ostream formatting; the tests only re-parse the
    # output, so any round-trippable decimal representation is acceptable.
    text = repr(value)
    return text


def _write_array(stream, array, indent):
    if not array:
        stream.write("{}")
        return
    stream.write("{\n")
    for i, item in enumerate(array):
        _write_indent(stream, indent + 1)
        _write_value(stream, item, indent + 1)
        if i + 1 < len(array):
            stream.write(",")
        stream.write("\n")
    _write_indent(stream, indent)
    stream.write("}")


def _write_class(stream, cls, indent):
    _write_indent(stream, indent)
    stream.write("class " + cls.name)
    if cls.parent:
        stream.write(": " + cls.parent)
    stream.write(" {\n")
    for prop in cls.properties:
        _write_indent(stream, indent + 1)
        stream.write(prop.name)
        if prop.value.is_array():
            stream.write("[]")
        stream.write(" += " if prop.extends else " = ")
        _write_value(stream, prop.value, indent + 1)
        stream.write(";\n")
    for child in cls.classes:
        _write_class(stream, child, indent + 1)
    _write_indent(stream, indent)
    stream.write("};\n")


# -- top-level config ----------------------------------------------------------
class Config:
    def __init__(self):
        self.root = Class("__root__")

    @staticmethod
    def read_file(path):
        with open(path, "r", encoding="latin-1") as handle:
            text = handle.read()
        parser = _Parser(_tokenize(text))
        config = Config()
        config.root = parser.parse_root()
        return config

    def write_file(self, path):
        with open(path, "w", encoding="latin-1", newline="") as handle:
            for cls in self.root.classes:
                _write_class(handle, cls, 0)
                handle.write("\n")

    def write_string(self):
        buffer = io.StringIO()
        for cls in self.root.classes:
            _write_class(buffer, cls, 0)
            buffer.write("\n")
        return buffer.getvalue()

    def skeletons(self):
        result = []
        cfg_skeletons = self.root.find_class("CfgSkeletons")
        if cfg_skeletons is None:
            return result
        for skeleton_class in cfg_skeletons.classes:
            skeleton = Skeleton(skeleton_class.name)
            bones = skeleton_class.find_property("skeletonBones")
            if bones is not None and bones.value.is_array():
                values = bones.value.as_array()
                i = 0
                while i + 1 < len(values):
                    if values[i].is_string() and values[i + 1].is_string():
                        skeleton.bones.append(SkeletonBone(values[i].as_string(), values[i + 1].as_string()))
                    i += 2
            result.append(skeleton)
        return result

    @staticmethod
    def skeleton_config(skeleton):
        skeleton_class = Class(skeleton.name)
        skeleton_class.properties.append(Property("isDiscrete", Value(0), False))
        skeleton_class.properties.append(Property("skeletonInherit", Value(""), False))
        bones = []
        for bone in skeleton.bones:
            bones.append(Value(bone.name))
            bones.append(Value(bone.parent))
        skeleton_class.properties.append(Property("skeletonBones", Value(bones), False))

        cfg_skeletons = Class("CfgSkeletons")
        cfg_skeletons.classes.append(skeleton_class)

        config = Config()
        config.root = Class("__root__")
        config.root.classes.append(cfg_skeletons)
        return config
